main.cpp

```diff
@@ -179,147 +134,532 @@ int main()
-    float divide(float lhs, float rhs);
+    float* value;
+
+    FloatType(float input);
+    ~FloatType();
+
+    FloatType& add(float input);
+    FloatType& add(const FloatType& input);
+    FloatType& add(const DoubleType& input);
+    FloatType& add(const IntType& input);
+    FloatType& subtract(float input);
+    FloatType& subtract(const FloatType& input);
+    FloatType& subtract(const DoubleType& input);
+    FloatType& subtract(const IntType& input);
+    FloatType& multiply(float input);
+    FloatType& multiply(const FloatType& input);
+    FloatType& multiply(const DoubleType& input);
+    FloatType& multiply(const IntType& input);
+    FloatType& divide(float input);
+    FloatType& divide(const FloatType& input);
+    FloatType& divide(const DoubleType& input);
+    FloatType& divide(const IntType& input);
 };
 
 struct DoubleType
 {
-    double add(double lhs, double rhs);
-    double subtract(double lhs, double rhs);
-    double multiply(double lhs, double rhs);
-    double divide(double lhs, double rhs);
+    double* value;
+
+    DoubleType(double input);
+    ~DoubleType();
+
+    DoubleType& add(double input);
+    DoubleType& add(const DoubleType& input);
+    DoubleType& add(const FloatType& input);
+    DoubleType& add(const IntType& input);
+    DoubleType& subtract(double input);
+    DoubleType& subtract(const DoubleType& input);
+    DoubleType& subtract(const FloatType& input);
+    DoubleType& subtract(const IntType& input);
+    DoubleType& multiply(double input);
+    DoubleType& multiply(const DoubleType& input);
+    DoubleType& multiply(const FloatType& input);
+    DoubleType& multiply(const IntType& input);
+    DoubleType& divide(double input);
+    DoubleType& divide(const DoubleType& input);
+    DoubleType& divide(const FloatType& input);
+    DoubleType& divide(const IntType& input);
 };
 
 struct IntType
 {
-    int add(int lhs, int rhs);
-    int subtract(int lhs, int rhs);
-    int multiply(int lhs, int rhs);
-    int divide(int lhs, int rhs);
+    int* value;
+
+    IntType(int input);
+    ~IntType();
+
+    IntType& add(int input);
+    IntType& add(const IntType& input);
+    IntType& add(const FloatType& input);
+    IntType& add(const DoubleType& input);
+    IntType& subtract(int input);
+    IntType& subtract(const IntType& input);
+    IntType& subtract(const FloatType& input);
+    IntType& subtract(const DoubleType& input);
+    IntType& multiply(int input);
+    IntType& multiply(const IntType& input);
+    IntType& multiply(const FloatType& input);
+    IntType& multiply(const DoubleType& input);
+    IntType& divide(int input);
+    IntType& divide(const IntType& input);
+    IntType& divide(const FloatType& input);
+    IntType& divide(const DoubleType& input);
 };
 
 // =========FLOATTYPE FUNCTION DEFINITIONS==================
 
-float FloatType::add(float lhs, float rhs)
+FloatType::FloatType(float input)
+{
+    value = new float(input);
+}
```

Comment:
When you have to choose where to initialize your member variables and your choices are:
- Constructor Member Initializer List
- Inside the Constructor's body

Try to use the Constructor Member Initializer List.
The reason is that the member variables need to be instantiated/initialized when the class is created.
When you assign to them in the constructor body, you are potentially re-initializing them, which is a waste of resources.
Take a look at the output of this program: https://godbolt.org/z/q83YM63ze
Take note of the second instance being created and destroyed due to that constructor body assignment.
- [ ] revise, use the member initializer list to initialize your member variables